File: cmd/cozysoc/https_check.go
package main

import (
	"context"
	"errors"
	"flag"
	"fmt"
	"os"
	"strings"
	"time"

	"github.com/fijimunkii/cozysoc/internal/controller/api"
	"github.com/fijimunkii/cozysoc/internal/controller/localapi"
)

// This is presentation over the existing typed client, not a second executor.
// The terminal implementation is compiled; flags cannot replace it or consent.
type httpsCheckClient interface {
	CheckHTTPS(context.Context, string, func(context.Context, api.HTTPSCheckReview) (bool, error)) (api.HTTPSCheckResult, error)
}

func runHTTPSCheckCommand(ctx context.Context, args []string, stdout, stderr *os.File) (err error) {
	fs := flag.NewFlagSet("https-check", flag.ContinueOnError)
	fs.SetOutput(stderr)
	stateDir := fs.String("state-dir", "", "controller state directory")
	fs.Usage = func() {
		fmt.Fprintln(stderr, "Usage: cozysoc https-check [--state-dir PATH] SELECTION_ID\nExperimental macOS one-shot HTTPS consent client. Requires a foreground terminal. Requires a controller started with --experimental-https-checks. Default: decline. No --yes, unattended mode or automatic retry.")
	}
	if err := fs.Parse(args); err != nil {
		if errors.Is(err, flag.ErrHelp) {
			return nil
		}
		return err
	}
	if fs.NArg() != 1 {
		return errors.New("https-check requires one opaque SELECTION_ID; flags must precede the target")
	}
	target := fs.Arg(0)
	if !localapi.ValidHTTPSSelectionID(target) {
		return errors.New("https-check requires a saved selection ID")
	}
	if err := ctx.Err(); err != nil {
		return err
	}
	// Refuse noninteractive use before loading controller credentials or dialing.
	terminal, err := openGatewayTerminal(os.Stdin, stdout)
	if err != nil {
		return errors.New("https-check requires normal-user foreground macOS terminal input and output; no controller connection was opened")
	}
	defer func() {
		if closeErr := terminal.Close(); closeErr != nil && err == nil {
			err = errors.New("terminal cleanup could not be confirmed; do not automatically retry a check")
		}
	}()
	dir, err := resolveStateDir(*stateDir)
	if err != nil {
		return err
	}
	return performHTTPSCheck(ctx, localapi.NewClient(dir), terminal, target)
}

func performHTTPSCheck(ctx context.Context, client httpsCheckClient, terminal gatewayCheckTerminal, target string) error {
	approved := false
	result, err := client.CheckHTTPS(ctx, target, func(decisionCtx context.Context, review api.HTTPSCheckReview) (bool, error) {
		if err := gatewayPromptContext(decisionCtx); err != nil {
			return false, err
		}
		if err := terminal.Write(decisionCtx, httpsReviewText(review)); err != nil {
			return false, err
		}
		// Discard type-ahead, including incomplete canonical lines, AFTER the
		// disclosure and BEFORE showing the exact approval prompt. No raw mode.
		if err := terminal.FlushInput(); err != nil {
			return false, err
		}
		prompt := fmt.Sprintf("Type \"check %s\" to authorize ONLY this sample. Anything else declines.\nApproval (default: decline): ", review.SelectionID)
		if err := terminal.Write(decisionCtx, prompt); err != nil {
			return false, err
		}
		line, err := terminal.ReadLine(decisionCtx)
		if err != nil {
			return false, err
		}
		if err := gatewayPromptContext(decisionCtx); err != nil {
			return false, err
		}
		if line != "check "+review.SelectionID+"\n" {
			return false, nil
		}
		// An output failure must still fail before consent, not send invisibly.
		if err := terminal.Write(decisionCtx, "Requesting the reviewed one-shot sample; no automatic retry.\n"); err != nil {
			return false, err
		}
		if err := gatewayPromptContext(decisionCtx); err != nil {
			return false, err
		}
		approved = true
		return true, nil
	})
	if err != nil {
		return httpsCommandError(err, approved)
	}
	outputCtx, cancel := context.WithTimeout(ctx, 2*time.Second)
	defer cancel()
	if err := terminal.Write(outputCtx, httpsResultText(result)); err != nil {
		return httpsCommandError(err, approved)
	}
	if result.Outcome != "completed" && result.Outcome != "declined" {
		return errors.New("https run did not complete normally; see the measured evidence above; do not automatically retry")
	}
	return nil
}

func httpsCommandError(err error, approved bool) error {
	if approved {
		return errors.New("https check outcome may be unknown; a request may have been sent; approval is consumed; do not automatically retry")
	}
	var response *localapi.ResponseError
	if errors.As(err, &response) && response.Code == "unavailable" {
		return errors.New("HTTPS checks require a separately started macOS controller with --experimental-https-checks and an active saved selection")
	}
	return errors.New("https review or terminal exchange failed; no approval was submitted; do not automatically retry")
}

// The typed client validates all values before presentation. Quoting the exact
// request makes its CRLF and any sensitive target characters visible as data.
func httpsReviewText(r api.HTTPSCheckReview) string {
	p, b, tls := r.Settings, r.Budget, r.Policy
	return fmt.Sprintf(`
EXPERIMENTAL: one-shot selected HTTPS check
Selection: %s
Endpoint: %s (%s); TLS server identity: %s
Request: %s %s; expected status: %d
Exact request bytes (quoted): %q
Destination policy: %s; outside enrolled prefixes: %t
Source: %s
Interface: %s (index %d); enrolled routable prefixes: %s
Profile: %s
Created: %s
Route observed: %s; route fresh until: %s
Review expires: %s (typing or resuming does not extend approval)

TLS: %s to %s; trust: %s; verify identity: %t; ALPN: %s; HTTP: %s
Fresh connection: %t; client authentication: %t; session resumption: %t; early data: %t
Proxy: %t; name resolution: %t; redirects: %t; response-body reading: %t
Limits: %d connection, %d request, %d retries; %d request bytes; %d response-header bytes.
TLS stream limits: %d read bytes / %d calls; %d write bytes / %d calls.
Deadlines: connect %d ms; TLS %d ms; response headers %d ms; total %d ms.
Concurrency: %d run; at least %d ms between admissions per controller.

Privacy:
%s

Private settings remain in local SQLite. Audits retain immutable references,
interface, times and normalized outcomes, not raw requests, response content or
TLS identities. This profile adds no cookies, authorization headers or request body.
A received response may be an error or redirect. One target failure does not prove
an internet outage or a security finding. This is one observer and one endpoint.
Same-binding network reuse cannot be distinguished by interface/prefix matching.
Physical Wi-Fi/NICs, VPNs, packaged permissions and sleep/resume are not certified.
Use only a network and endpoint you may test.

`, r.SelectionID, p.Endpoint, p.Family, p.ServerName, p.Method, p.RequestTarget, p.ExpectedStatus, r.RequestBytes, p.DestinationPolicy, r.OutsideEnrolledPrefixes, r.Source, r.Binding.InterfaceName, r.Binding.InterfaceIndex, strings.Join(r.Binding.Prefixes, ", "), r.Profile, r.CreatedAt.Format(time.RFC3339Nano), r.RouteObservedAt.Format(time.RFC3339Nano), r.RouteFreshUntil.Format(time.RFC3339Nano), r.ExpiresAt.Format(time.RFC3339Nano), tls.MinTLSVersion, tls.MaxTLSVersion, tls.TrustStore, tls.VerifyServerIdentity, tls.ALPN, tls.HTTPVersion, tls.FreshConnection, tls.ClientAuthentication, tls.SessionResumption, tls.EarlyData, tls.UseProxy, tls.ResolveNames, tls.FollowRedirects, tls.ReadResponseBody, b.MaxConnections, b.MaxRequests, b.MaxRetries, b.MaxRequestBytes, b.MaxResponseHeaderBytes, b.MaxTransportReadBytes, b.MaxTransportReadCalls, b.MaxTransportWriteBytes, b.MaxTransportWriteCalls, b.ConnectTimeoutMS, b.TLSHandshakeTimeoutMS, b.ResponseHeaderTimeoutMS, b.TotalTimeoutMS, b.MaxConcurrentRuns, b.MinRunIntervalMS, strings.Join(r.Privacy, "\n"))
}

func httpsResultText(r api.HTTPSCheckResult) string {
	if r.Outcome == "declined" {
		return "Declined. No HTTPS request was authorized.\n"
	}
	text := fmt.Sprintf("Run: %s\nExecution outcome: %s\n", r.RunID, r.Outcome)
	if m := r.Measurement; m != nil {
		text += fmt.Sprintf("Historical sample: %s to %s\nStage: %s; request: %s; HTTPS exchange: %s\n", m.StartedAt.Format(time.RFC3339Nano), m.CompletedAt.Format(time.RFC3339Nano), m.Stage, m.Request, m.Exchange)
		if m.StatusCode != 0 {
			text += fmt.Sprintf("Received HTTP status: %d; expected: %d; matches expectation: %t\n", m.StatusCode, r.Review.Settings.ExpectedStatus, m.StatusCode == r.Review.Settings.ExpectedStatus)
		}
		if m.ResponseTimeNanoseconds != nil {
			text += fmt.Sprintf("Time to final response header: %s (includes connect/TLS; not pure network RTT)\n", time.Duration(*m.ResponseTimeNanoseconds))
		}
	} else {
		text += "No measurement was confirmed.\n"
	}
	return text + "Historical evidence only. A single endpoint response or timeout does not establish internet availability or security. Do not automatically retry.\n"
}

File: ui/src/quality/https-check.ts
export interface HTTPSSettings {
  endpoint: string; server_name: string; request_target: string;
  family: "ipv4" | "ipv6"; method: "HEAD" | "GET"; expected_status: number;
  destination_policy: "exact-endpoint";
}
export interface HTTPSSelection { selection_id: string; settings: HTTPSSettings }
export interface HTTPSCheckReview extends HTTPSSelection {
  review_id: string; created_at: string; expires_at: string; source: string;
  interface_name: string; interface_index: number; prefixes: string[];
  outside_enrolled_prefixes: boolean; route_observed_at: string; route_fresh_until: string;
  policy: HTTPSPolicy; request_bytes: string; privacy: string[]; budget: HTTPSBudget;
}
export interface HTTPSPolicy {
  alpn: string; trust_store: string; client_authentication: boolean; http_version: string;
  min_tls_version: string; max_tls_version: string; verify_server_identity: boolean;
  fresh_connection: boolean; session_resumption: boolean; early_data: boolean;
  use_proxy: boolean; resolve_names: boolean; follow_redirects: boolean; read_response_body: boolean;
}
export interface HTTPSBudget {
  max_connections:number; max_requests:number; max_retries:number; max_request_bytes:number;
  max_response_header_bytes:number; max_transport_read_bytes:number; max_transport_write_bytes:number;
  max_transport_read_calls:number; max_transport_write_calls:number;
  connect_timeout_ms:number; tls_handshake_timeout_ms:number; response_header_timeout_ms:number;
  total_timeout_ms:number; max_concurrent_runs:number; min_run_interval_ms:number;
}
export interface HTTPSCheckResult { outcome:"declined"|"completed"|"failed"|"canceled"|"indeterminate"|"blocked"; run_id?:string; failure_code?:string }
export interface HTTPSCheckClient { reviewHTTPS(selectionID:string):Promise<HTTPSCheckReview>; decideHTTPS(reviewID:string,approve:boolean):Promise<HTTPSCheckResult> }
const selectionID=/^https-selection\.[0-9a-f]{32}$/;
function invalid():never {throw new Error("HTTPS check response is invalid.")}
function exact(raw:unknown,required:string[],optional:string[]=[]):Record<string,unknown>{
  if(raw===null||typeof raw!=="object"||Array.isArray(raw))return invalid();
  const v=raw as Record<string,unknown>;
  if(required.some((k)=>!Object.hasOwn(v,k))||Object.keys(v).some((k)=>!required.includes(k)&&!optional.includes(k)))return invalid();
  return v;
}
function int(raw:unknown,low:number,high:number):number {if(typeof raw!=="number"||!Number.isSafeInteger(raw)||raw<low||raw>high)return invalid();return raw}
function time(raw:unknown):string {if(typeof raw!=="string"||!/^\d{4}-\d\d-\d\dT\d\d:\d\d:\d\d(?:\.\d{1,9})?Z$/.test(raw)||!Number.isFinite(Date.parse(raw)))return invalid();return raw}
function settings(raw:unknown):HTTPSSettings{
  const v=exact(raw,["endpoint","server_name","request_target","family","method","expected_status","destination_policy"]);
  if(typeof v.endpoint!=="string"||v.endpoint.length>64||!/^(?:\[[0-9a-f:]+\]|[0-9.]+):443$/.test(v.endpoint)||
    typeof v.server_name!=="string"||v.server_name.length>253||!/^[-a-z0-9.]+$/.test(v.server_name)||
    typeof v.request_target!=="string"||v.request_target.length<1||v.request_target.length>1024||!v.request_target.startsWith("/")||v.request_target.startsWith("//")||
    /[\u0000-\u001f\u007f\\]/.test(v.request_target)||
    (v.family!=="ipv4"&&v.family!=="ipv6")||(v.method!=="HEAD"&&v.method!=="GET")||int(v.expected_status,100,599)<100||v.destination_policy!=="exact-endpoint")return invalid();
  return v as unknown as HTTPSSettings;
}
function selection(raw:unknown):HTTPSSelection{
  const v=exact(raw,["selection_id","settings"]);
  if(typeof v.selection_id!=="string"||!selectionID.test(v.selection_id))return invalid();
  return {selection_id:v.selection_id,settings:settings(v.settings)};
}
export function parseHTTPSSelections(raw:unknown):HTTPSSelection[]{
  const v=exact(raw,["items"]);if(!Array.isArray(v.items)||v.items.length>16)return invalid();
  const items=v.items.map(selection);if(new Set(items.map((i)=>i.selection_id)).size!==items.length)return invalid();return items;
}
function policy(raw:unknown):HTTPSPolicy{
  const keys=["alpn","trust_store","client_authentication","http_version","min_tls_version","max_tls_version","verify_server_identity","fresh_connection","session_resumption","early_data","use_proxy","resolve_names","follow_redirects","read_response_body"];
  const v=exact(raw,keys);
  const expected:HTTPSPolicy={alpn:"http/1.1",trust_store:"system",client_authentication:false,http_version:"HTTP/1.1",min_tls_version:"TLS 1.2",max_tls_version:"TLS 1.3",verify_server_identity:true,fresh_connection:true,session_resumption:false,early_data:false,use_proxy:false,resolve_names:false,follow_redirects:false,read_response_body:false};
  if(keys.some((k)=>v[k]!==expected[k as keyof HTTPSPolicy]))return invalid();return expected;
}
function budget(raw:unknown):HTTPSBudget{
  const keys=["max_connections","max_requests","max_retries","max_request_bytes","max_response_header_bytes","max_transport_read_bytes","max_transport_write_bytes","max_transport_read_calls","max_transport_write_calls","connect_timeout_ms","tls_handshake_timeout_ms","response_header_timeout_ms","total_timeout_ms","max_concurrent_runs","min_run_interval_ms"];
  const v=exact(raw,keys);
  const fixed:Omit<HTTPSBudget,"max_request_bytes">={max_connections:1,max_requests:1,max_retries:0,max_response_header_bytes:16384,max_transport_read_bytes:131072,max_transport_write_bytes:32768,max_transport_read_calls:512,max_transport_write_calls:64,connect_timeout_ms:2000,tls_handshake_timeout_ms:3000,response_header_timeout_ms:2000,total_timeout_ms:8000,max_concurrent_runs:1,min_run_interval_ms:60000};
  if(Object.keys(fixed).some((k)=>v[k]!==fixed[k as keyof typeof fixed]))return invalid();
  return {...fixed,max_request_bytes:int(v.max_request_bytes,1,4096)};
}
export function parseHTTPSReview(raw:unknown):HTTPSCheckReview{
  const v=exact(raw,["review_id","selection_id","settings","created_at","expires_at","source","interface_name","interface_index","prefixes","outside_enrolled_prefixes","route_observed_at","route_fresh_until","policy","request_bytes","privacy","budget"]);
  if(typeof v.review_id!=="string"||!/^[A-Za-z0-9_-]{43}$/.test(v.review_id)||typeof v.selection_id!=="string"||!selectionID.test(v.selection_id)||
    typeof v.source!=="string"||v.source.length>64||!/^[0-9a-f:.]+$/.test(v.source)||typeof v.interface_name!=="string"||!/^[A-Za-z0-9][A-Za-z0-9_.:-]{0,63}$/.test(v.interface_name)||
    !Array.isArray(v.prefixes)||v.prefixes.length<1||v.prefixes.length>64||v.prefixes.some((p)=>typeof p!=="string"||p.length>128||!/^[0-9a-f:./]+$/.test(p))||
    typeof v.outside_enrolled_prefixes!=="boolean"||typeof v.request_bytes!=="string"||v.request_bytes.length>4096||!Array.isArray(v.privacy)||v.privacy.length!==4||v.privacy.some((p)=>typeof p!=="string"||p.length<1||p.length>600))return invalid();
  const created=time(v.created_at),expires=time(v.expires_at),observed=time(v.route_observed_at),fresh=time(v.route_fresh_until);
  if(Date.parse(expires)-Date.parse(created)!==30000||Date.parse(fresh)-Date.parse(observed)!==30000||Date.parse(observed)>Date.parse(created))return invalid();
  const s=settings(v.settings),b=budget(v.budget);
  if(!v.request_bytes.startsWith(`${s.method} ${s.request_target} HTTP/1.1\r\n`)||!v.request_bytes.includes(`\r\nHost: ${s.server_name}\r\n`)||b.max_request_bytes!==new TextEncoder().encode(v.request_bytes).length)return invalid();
  return {review_id:v.review_id,selection_id:v.selection_id,settings:s,created_at:created,expires_at:expires,source:v.source,interface_name:v.interface_name,interface_index:int(v.interface_index,1,2147483647),
    prefixes:v.prefixes as string[],outside_enrolled_prefixes:v.outside_enrolled_prefixes,route_observed_at:observed,route_fresh_until:fresh,policy:policy(v.policy),request_bytes:v.request_bytes,privacy:v.privacy as string[],budget:b};
}
export function parseHTTPSResult(raw:unknown):HTTPSCheckResult{
  const v=exact(raw,["outcome"],["run_id","failure_code"]);
  if(v.outcome==="declined"){if(v.run_id!==undefined||v.failure_code!==undefined)return invalid();return {outcome:"declined"}}
  if(!["completed","failed","canceled","indeterminate","blocked"].includes(String(v.outcome))||typeof v.run_id!=="string"||!/^[0-9a-f]{32}$/.test(v.run_id))return invalid();
  const expected:Record<string,string[]>={completed:[""],failed:["execution_failed"],canceled:["canceled"],indeterminate:["execution_failed"],blocked:["precondition_failed","review_expired"]};
  const failure=v.failure_code===undefined?"":v.failure_code;
  if(typeof failure!=="string"||!expected[String(v.outcome)]?.includes(failure))return invalid();
  return {outcome:v.outcome as HTTPSCheckResult["outcome"],run_id:v.run_id,...(failure?{failure_code:failure}:{})};
}
